### PR TITLE
ENH: Extend the nonlinear transforms API

### PR DESCRIPTION
**Pull request by oesteban**
This PR lays the ground for future work on #56, and #89, by defining the matrix multiplication operator on field-based transforms.

### CODE REVIEW

**Review by effigies — COMMENTED** [review]
Conceptually, a `DisplacementsFieldTransform` is not a `DeformationFieldTransform`. A `DeformationFieldTransform` has a coherent `__matmul__` implementation, while that same implementation does not make sense for `DisplacementsFieldTransform`, so this inheritance is bound to cause confusion.

I think we're mixing up composition and application and calling both of them "transforms". Your `DeformationFieldTransform._field` is equivalent to `Affine._matrix`, while your `DisplacementsFieldTransform._field` is equivalent to having already applied `Affine.map()` to the voxel indices, turning it into a lookup table.

Maybe we need to think about a typing structure like:

```
class Composable:
    def __matmul__(self, other): ...

class Applicable:
    def map(self, points): ...
```

Deformation fields and reference-less affines are composable, while displacement fields and affines with references are applicable.

**Comment by oesteban** [comment]
One more reason to use the proposed design is #155. With the current implementation, that one becomes trivial for both types of fields.

**Comment by effigies** [comment]
My understanding here is that `DeformationFieldTransform` has an identity during composition and application, while `DisplacementsFieldTransform` can have an identity during application (and it's essentially the same), but that identity does not hold during composition.

I have not tested the following, but here's my rough expectation from reading this code:
```Python
>>> dfm = DeformationFieldTransform()
>>> disp = DisplacementsFieldTransform()
```

Application

```Python
>>> x, y, z = some_coord
>>> np.array_equal(dfm.map([x, y, z]), [x, y, z]))
True
>>> np.array_equal(disp.map([x, y, z]), [x, y, z]))
True
```

Composition

```Python
>>> dfm @ dfm == dfm
True
>>> disp @ disp == disp
False
```

**Comment by effigies** [comment]
Maybe a better way to put this is that a `DeformationFieldTransform` seems like the correct internal representation, and that we should not have a `DisplacementsFieldTransform` except possibly as an implementation detail for certain formats.

**Comment by oesteban** [comment]
Okay, I think I now understand what you mean.

If you assume that:
```Python
>>> dfm = DeformationFieldTransform()
>>> disp = DisplacementsFieldTransform()
```
will create identity transforms for both structures then:

* `dfm._field` contains an $(S_x, S_y, S_z, D)$ array where each voxel contains the three coordinates of that voxel in physical space.
* `disp._displacements` is all zeros, and `disp._field == dfm._field`.

Then, mapping is equivalent for both.

Regarding composition, in both cases:

```Python
>>> dfm @ dfm == dfm
True
>>> disp @ disp == disp
True
```
iff they are identity.

For both of them, if they are not identity, then both satisfy:
```Python
>>> dfm @ dfm == dfm
False
>>> disp @ disp == disp
False
```
which will become more apparent with the implementation of #155. In the case of `disp` I think you were already hinting at this, but it is exactly the same for `dfm`.

Perhaps you would feel more comfortable with a single representation structure:

```Python
class DenseFieldTransform(TransformBase):
    def __init__(self, deformation=None, displacements=None, reference=None):
        ...
```

where:
1. deformation and displacements are mutually exclusive
2. deformation (`self._field`) or displacements (`self._displacements`) are calculated from the provided argument
3. share the rest of the implementation.
4. if both deformation and displacements are `None`, an identity field is generated.

Does this make sense?

File: nitransforms/base.py [hunk 1]
```diff
@@ -303,6 +303,10 @@ def apply(
 
         return resampled
 
+    def __matmul__(self, b):
+        """Compose with a transform on the right."""
+        return b
```
[the remark this answers is not in the document]
Comment (oesteban) [reply nitransforms/base.py:308]: Okay, I'm not sure of the usefulness of one more layer of interfacing, but I think this particular change can be pushed away to some future moment where we make a specific decision on this.

**Comment by effigies** [comment]
Okay, reading the new API, I think some things that were unclear are clicking. Now I'm just left to wonder: Why do we want to have the `_displacements` pre-calculated at all? Why do we not just use:

```Python
class DenseFieldTransform(TransformBase):
    def __init__(self, deltas, reference=None):
        self._deltas = deltas
        self._reference = reference

    def __matmul__(self, other):
        deltas = b.map(
            self._field.reshape((-1, self._deltas.shape[-1]))
        ).reshape(self._deltas.shape)
        return DenseFieldTransform(deltas, reference=self.reference)

    def map(self, coords, inverse=False):
        indices = calculate_indices(coords)
        return coords + self._deltas[indices]

    @classmethod
    def from_deformation_field(cls, deformations, reference=None):
        if reference is None:
            reference = ...
        deltas = deformations - reference.ndcoords.T.reshape(deformations.shape)
        return cls(deltas, reference=reference)
```

Finally, are we sure that we want to use `__matmul__` and not `__rmatmul__`? We would need to think of a transformation that would be both simple and obviously not commutative to verify that this goes in the direction we expect, but I can't immediately think of one.

Apologies if I'm jumping all over the place. Trying to understand this somewhat quickly and may be flailing a bit...

**Comment by oesteban** [comment]
> Why do we want to have the `_displacements` pre-calculated at all?

I think this is just for reproducibility purposes -- to be able to write it out to disk (possibly in a different format such as X5) without alteration of the array contents.

> Why do we not just use:

As commented in the code, AFAIK the deformations are numerically more stable and that's why I made them the default.

> Finally, are we sure that we want to use ``__matmul__`` and not ``__rmatmul__``? We would need to think of a transformation that would be both simple and obviously not commutative to verify that this goes in the direction we expect, but I can't immediately think of one.

I opened an issue to remind me of this good point.

> Apologies if I'm jumping all over the place. Trying to understand this somewhat quickly and may be flailing a bit...

On the contrary, this is very helpful. Thanks for your time.

I'll use your ``_deltas`` convention because I like it better than ``_displacements``. Other than that, I will call this PR finished and move on. Thanks a lot for the feedback. Happy to come back to any comments anytime.

**Pull request merged**